### ISSUE TITLE
Support for Microsoft.Data.SqlClient

### ISSUE THREAD

**Issue by DavidBoike** [Particular/NServiceBus.SqlServer#506]
Support for the [new Microsoft.Data.SqlClient](https://devblogs.microsoft.com/dotnet/introducing-the-new-microsoftdatasqlclient/) library in the SQL Transport.

Related issue: [support in SQL Persistence](https://github.com/Particular/NServiceBus.Persistence.Sql/issues/363)

**Comment by DavidBoike** [comment]
The first step here is for us to analyze how this will affect both SQL Transport and Persistence, as well as customers that are still using (and have no reason to switch from) the original SqlClient.

**Comment by SimonCropp** [comment]
Microsoft.Data.SqlClient  has now been out for 4months. any progress on this?

**Comment by robbaman** [comment]
Exactly, we're running into issues on all services that use EF Core 3.0 as the new SqlClient is the one used there. We can no longer have EF enlist in the Outbox transaction which is unacceptable for us.

Isn't it possible to 'just' create a new SqlDialect implementation for the new SqlClient and leave the old implementation untouched for maximum backwards compatibility?

**Comment by robbaman** [comment]
To expand on what I just posted, we're currently using code similar to what is described in the docs (https://docs.particular.net/persistence/sql/dialect-mssql):

```
var persistence = endpointConfiguration.UsePersistence<SqlPersistence>();
var subscriptions = persistence.SubscriptionSettings();
subscriptions.CacheFor(TimeSpan.FromMinutes(1));
persistence.SqlDialect<SqlDialect.MsSqlServer>();
persistence.ConnectionBuilder(
    connectionBuilder: () =>
    {
        return new SqlConnection(connection);
    });
```

The big difference being that 'our' SqlConnection is of the Microsoft.Data variant. Now this shouldn't be an issue because of the use of the `.CreateCommand()` and `.CreateParameter()` methods.

In the `MsSqlServer` dialect file however there is one explicit mention of the SqlParameter type in the code that works with the ArraySegment:

```
internal override void SetParameterValue(DbParameter parameter, object value)
{
    //TODO: do ArraySegment fro outbox
    if (value is ArraySegment<char> charSegment)
    {
        var sqlParameter = (SqlParameter)parameter;

        sqlParameter.Value = charSegment.Array;
        sqlParameter.Offset = charSegment.Offset;
        sqlParameter.Size = charSegment.Count;
    }
    else
    {
        parameter.Value = value;
    }
}
```

Note how the parameter is cast to the `SqlParameter` type, which in our case yields an InvalidCastException.

Since we are heavily impacted by this but, it'd be great if a solution could be provided in the short term.

Thanks!

**Comment by SimonCropp** [comment]
bump

**Comment by SimonCropp** [comment]
is this thing on?

**Comment by HEskandari** [comment]
We did evaluate Microsoft.Data.SqlClient. We're also in the process of checking if we can port SQL Persistence and Transport to use the base class/interfaces in System.Data.Common so we can move away from using concrete classes. Based on this result, we'll decide to either release a new version with the new SQLClient or the System.Data.Common package as the only dependency.

**Comment by SimonCropp** [comment]
@HEskandari how long will this decision take?

**Comment by HEskandari** [comment]
@SimonCropp We're at our company conference this week but this spike is the next thing we'll be working on. We'll keep you posted about the results.

**Comment by robbaman** [comment]
> We did evaluate Microsoft.Data.SqlClient. We're also in the process of checking if we can port SQL Persistence and Transport to use the base class/interfaces in System.Data.Common so we can move away from using concrete classes. Based on this result, we'll decide to either release a new version with the new SQLClient or the System.Data.Common package as the only dependency.

I'm a bit confused you mention you're looking into either 1) Move away from concrete classes or 2) use one of the two (but not both).

How is option 2) a solution? As long as only one of the two libraries is supported in some fashion there will be a group that won't be able to use this package. But perhaps I'm misinterpreting your response.

**Comment by HEskandari** [comment]
@robbaman let me rephrase:

- We have already spiked supporting the new Microsoft.Data.SqlClient. It is possible and we can adopt it.
- We think that it is a better long-term strategy to use the `System.Data.Common` and use the abstractions it provides.

If our Sql Persistence/Transport libraries expect the abstractions available in the `System.Data.Common` package, the users of our packages can use any library that implements those, which means you can use the new and the old SqlClient package as they both implement these interfaces. For us, this means we don't have to create/support/maintain more packages (one for the old SqlClient and one for the new SqlClient).

**Comment by robbaman** [comment]
Ah wait, I see now where my confusion comes from. You wrote about having `System.Data.Common` as the only dependency where I read `System.Data.SqlClient`. Apologies.

**Comment by markgould** [comment]
Also very interested in this. Porting an application to .NET Core 3.1 to find out it isn't working due to this issue.

Is there an ETA when this might be available?

**Comment by markgould** [comment]
I see that there was a commit made recently and a prerelease package pushed to MyGet. Referenced NServiceBus.Transport.SqlServer instead of NServiceBus.SqlServer and updated to latest pre-release of NServiceBus.Persistence.Sql.

Everything appears to be working - looking forward to the stable releases.

**Comment by bording** [comment]
@markgould As you've discovered, we are working on this now, so a final release should be happening sometime soon, but I don't have a specific ETA.

**Comment by bording** [comment]
NServiceBus.SqlServer and NServiceBus.Transport.SqlServer 6.0 have been released. 

NServiceBus.SqlServer continues to reference `System.Data.SqlClient`, and the new NServiceBus.Transport.SqlServer package references `Microsoft.Data.SqlClient`.
